core/src/main/java/org/apache/calcite/sql/fun/SqlBigQueryFormatDatetimeFunction.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.sql.fun;
+
+import org.apache.calcite.sql.SqlCallBinding;
+import org.apache.calcite.sql.SqlFunction;
+import org.apache.calcite.sql.SqlFunctionCategory;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlOperandCountRange;
+import org.apache.calcite.sql.SqlOperator;
+import org.apache.calcite.sql.type.ReturnTypes;
+import org.apache.calcite.sql.type.SqlOperandCountRanges;
+import org.apache.calcite.sql.type.SqlOperandTypeChecker;
+import org.apache.calcite.sql.type.SqlTypeFamily;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+import java.util.Locale;
+
+import static org.apache.calcite.sql.type.SqlTypeName.DATE;
+import static org.apache.calcite.sql.type.SqlTypeName.TIME;
+
+/**
+ * <p> The Google BigQuery style datetime formatting functions. This is a generic type representing
+ * one of the following:</p>
+ *
+ * <ul>
+ *   <li>{@code FORMAT_TIME(format_string, time_object)}
+ *   <a href="https://cloud.google.com/bigquery/docs/reference/standard-sql/time_functions#format_time">ref</a></li>
+ *   <li>{@code FORMAT_DATE(format_string, date_expr)}
+ *   <a href="https://cloud.google.com/bigquery/docs/reference/standard-sql/date_functions#format_date">ref</a></li>
+ *   <li>{@code FORMAT_TIMESTAMP(format_string, timestamp[, time_zone])}
+ *  <a href="https://cloud.google.com/bigquery/docs/reference/standard-sql/timestamp_functions#format_timestamp">ref</a></li>
+ *  <li>{@code FORMAT_DATETIME(format_string, timestamp[, time_zone])}
+ *  <a href="https://cloud.google.com/bigquery/docs/reference/standard-sql/datetime_functions#format_datetime">ref</a></li>
+ * </ul>
+ */
+public class SqlBigQueryFormatDatetimeFunction extends SqlFunction {
```

Comment:
Parsing/formatting functions might have enough commonality to warrant a common class - an extension to `SqlBasicFunction` with an extra field, a map from strings to format elements.

I wouldn't change it now. But come back and refactor if that pattern emerges.

Reply:
However I would make the constructor private now. A public constructor will prevent you from subclassing later. Use a `create` method instead, and add `withXxx` methods to change the value of parameters that are not frequently used. (Same pattern as `SqlBasicFunction`, and a pattern that is a little nearer to functional programming than object-oriented.)